akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -444,108 +568,173 @@ private[akka] class Shard(
+    case _: ShardRegion.StartEntityAck =>
+      // FIXME if we got an ack it was started elsewhere we should probably make sure we don't have it in current or pending starts?
+      stash()
+    case ShardRegion.Passivate(stopMessage) =>
+      if (VerboseDebug)
+        log.debug(
+          "Passivation of [{}] arrived while updating.",
+          entities.entityId(sender()).getOrElse(s"Unknown actor ${sender()}"))
+      passivate(sender(), stopMessage)
+    case _: ShardRegionCommand           => stash()
+    case msg: ShardQuery                 => receiveShardQuery(msg)
+    case PassivateIdleTick               => stash()
+    case msg: RememberEntityStoreCrashed => rememberEntityStoreCrashed(msg)
+    case msg if extractEntityId.isDefinedAt(msg) =>
+      deliverMessage(msg, sender())
+    case msg =>
+      // shouldn't be any other message types, but just in case
+      log.warning(
+        "Stashing unexpected message [{}] while waiting for remember entities update of starts [{}], stops [{}]",
+        msg.getClass,
+        update.started.mkString(", "),
+        update.stopped.mkString(", "))
+      stash()
+
+  }
+
+  def onUpdateDone(starts: Set[EntityId], stops: Set[EntityId]): Unit = {
+    // entities can contain both ids from start/stops and pending ones, so we need
+    // to mark the completed ones as complete to get the set of pending ones
+    starts.foreach { entityId =>
+      val stateBeforeStart = entities.entity(entityId).getOrElse(NoState)
+      // this will start the entity and transition the entity state in sessions to active
+      getOrCreateEntity(entityId)
+      sendMsgBuffer(entityId)
+      stateBeforeStart match {
+        case RememberingStart(Some(ackTo)) => ackTo ! ShardRegion.StartEntityAck(entityId, shardId)
+        case _                             =>
+      }
+      touchLastMessageTimestamp(entityId)
+    }
+    stops.foreach { entityId =>
+      // FIXME actually stop it by sendint stop message or what used to happen when stop write completed
+      entities.entityState(entityId) match {
+        case RememberingStop(Passivating) =>
+          // this updates entity state
+          passivateCompleted(entityId)
+        case RememberingStop(StartedElsewhere) =>
+          // Drop buffered messages if any (to not cause re-ordering)
+          messageBuffers.remove(entityId)
+          entities.removeEntity(entityId)
+        case state =>
+          throw new IllegalStateException(
+            s"Unexpected state [$state] when storing stop completed for entity id [$entityId]")
+      }
+    }
+
+    val (pendingStarts, pendingStops) = entities.pendingRememberEntities()
+    if (pendingStarts.isEmpty && pendingStops.isEmpty) {
+      if (VerboseDebug) log.debug("Update complete, no pending updates, going to idle")
+      unstashAll()
+      context.become(idle)
+    } else {
+      // FIXME no unstashing as long as we are batching, is that a problem?
```

Comment:
There's a certain unfairness about it but i think it is okay as long as we try and handle most messages in this state, there's a few that we stash that we can act on right away